src/libs/elektra/diff.c

```diff
@@ -530,11 +932,33 @@ KeySet * elektraDiffGetModifiedMetaKeys (const ElektraDiff * ksd, Key * key)
 
 	KeySet * modifiedKeys = ksNew (0, KS_END);
 
-	findDifferences (keyMeta (key), keyMeta (oldKey), NULL, NULL, modifiedKeys, NULL);
+	findDifferences (keyMeta (key), keyMeta (oldKey), NULL, NULL, modifiedKeys, NULL, NULL);
 
 	return modifiedKeys;
 }
 
+/**
+ * Undos the changes that are presented in the diff.
+ *  - Added keys get removed from @p ks
+ *  - Removed keys get added back to @p ks,
+ *  - Modified keys will get their old values and meta data in @p ks
+ *
+ * @param ksd the diff
+ * @param ks the keyset that shall be undone
+ */
+void elektraDiffUndo (ElektraDiff * ksd, KeySet * ks)
+{
+	for (elektraCursor i = 0; i < ksGetSize (ksd->addedKeys); i++)
+	{
+		Key * toRemove = ksAtCursor (ksd->addedKeys, i);
+		Key * key = ksLookup (ks, toRemove, KDB_O_POP);
+		keyDel (key);
+	}
```

Comment:
Done! It's about 10% improvement.

Reply:
That is more than I expected. Although it probably depends on the test data. I assume 10% will be if it was mostly `ksLookup`s with binary search and with not the hashmap.

Reply:
Well that's probably because we're not actually using the hashmap by default. You'd have to put an extra flag into `ksLookup` to do so. Which is good, because in my tests with a very high number of keys, the cost of building the hashmap **far** outweighs any gains in lookup time. We're talking about a factor of 5 or more here ...

Reply:
Yeah, I know... There is some sort of semi-smart prediction logic to determine whether the hashmap gets built.